File: 2/BalancedBrackets.java
import java.util.*;

class BalancedBrackets {
	public static boolean hasBalancedBrackets(String brackets) {
		Map<Character, Character> bracketsPair = Map.of('(', ')', '{', '}', '[', ']');
		
		var stack = new Stack<Character>();
		for(int i=0; i<brackets.length(); i++) {
			if(stack.isEmpty()) {
				if(bracketsPair.get(brackets.charAt(i)) != null) {
					stack.push(brackets.charAt(i));
				} else {
					return false;
				}
			} else {
				if(bracketsPair.get(brackets.charAt(i)) == null) {
					if(bracketsPair.get(stack.peek()) == brackets.charAt(i)) {
						stack.pop();
					} else {
						return false;
					}
				} else {
						stack.push(brackets.charAt(i));
				}
			}
		}


		return true;
	}
	
	public static void main(String args[]) {
		String brackets = "(([]()()){})";	
		System.out.println(hasBalancedBrackets(brackets));
	}
}

File: 2/js/src/BalancedBrackets.ts
function isBalanced(brackets: string): boolean {
	const map: Map<String, String> = new Map();
	map.set("(", ")");
	map.set("[", "]");
	map.set("{", "}");

	const stack: string[] = [];
	for(const bracket of brackets) {
		if(map.has(bracket)) {
			stack.push(bracket);
		} else {
			if(map.get(stack.pop()!) != bracket) {
				return false;
			}
		}
	}

	return stack.length == 0;
}

(function main() {
	const brackets = "(([]()()){})";
	console.log(isBalanced(brackets))
})();
